File: frontend/lib/accountService.ts
import executionApi from "./executionApi";

export interface AccountResponse {
  id: string;
  userId: string;
  role: string;
  accountNumber: string;
  balance: number;
  pendingBalance: number;
  createdAt: string;
}

export interface BankAccountResponse {
  id: string;
  bankName: string;
  accountNumber: string;
  accountHolderName: string;
  defaultAccount: boolean;
  createdAt: string;
  updatedAt: string;
}

export interface BankAccountRequest {
  bankName: string;
  accountNumber: string;
  accountHolderName: string;
}

export interface InitiateDepositResponse {
  transactionId: string;
  referenceId: string;
  status: string;
}

export interface DepositStatusResponse {
  referenceId: string;
  status: string;
  failureReason?: string;
  account?: AccountResponse;
}

export const accountService = {
  async getAccount(): Promise<AccountResponse> {
    const res = await executionApi.get<AccountResponse>("/api/account");
    return res.data;
  },

  async initiateDeposit(projectId: string, amount: number, payerMsisdn: string): Promise<InitiateDepositResponse> {
    const res = await executionApi.post<InitiateDepositResponse>("/api/account/deposit/initiate", {
      projectId,
      amount,
      payerMsisdn,
    });
    return res.data;
  },

  async getDepositStatus(referenceId: string): Promise<DepositStatusResponse> {
    const res = await executionApi.get<DepositStatusResponse>(`/api/account/deposit/${referenceId}/status`);
    return res.data;
  },

  async getBankAccounts(): Promise<BankAccountResponse[]> {
    const res = await executionApi.get<BankAccountResponse[]>("/api/account/bank-accounts");
    return res.data;
  },

  async addBankAccount(data: BankAccountRequest): Promise<BankAccountResponse> {
    const res = await executionApi.post<BankAccountResponse>("/api/account/bank-accounts", data);
    return res.data;
  },

  async updateBankAccount(id: string, data: BankAccountRequest): Promise<BankAccountResponse> {
    const res = await executionApi.put<BankAccountResponse>(`/api/account/bank-accounts/${id}`, data);
    return res.data;
  },

  async deleteBankAccount(id: string): Promise<void> {
    await executionApi.delete(`/api/account/bank-accounts/${id}`);
  },

  async setDefaultBankAccount(id: string): Promise<void> {
    await executionApi.patch(`/api/account/bank-accounts/${id}/default`);
  },
};

File: backend/ProjectWorker-Execution-Service/src/main/java/com/example/ProjectWorker_Execution_Service/service/AccountService.java
package com.example.ProjectWorker_Execution_Service.service;

import com.example.ProjectWorker_Execution_Service.dto.AccountResponse;
import com.example.ProjectWorker_Execution_Service.dto.BankAccountRequest;
import com.example.ProjectWorker_Execution_Service.dto.BankAccountResponse;
import com.example.ProjectWorker_Execution_Service.dto.DepositStatusResponse;
import com.example.ProjectWorker_Execution_Service.dto.InitiateDepositRequest;
import com.example.ProjectWorker_Execution_Service.dto.InitiateDepositResponse;
import com.example.ProjectWorker_Execution_Service.security.UserPrincipal;

import java.util.List;

public interface AccountService {

    AccountResponse getOrCreateAccount(UserPrincipal principal);

    InitiateDepositResponse initiateDeposit(InitiateDepositRequest request, UserPrincipal principal);

    DepositStatusResponse getDepositStatus(String referenceId, UserPrincipal principal);

    List<BankAccountResponse> getBankAccounts(UserPrincipal principal);

    BankAccountResponse addBankAccount(BankAccountRequest request, UserPrincipal principal);

    BankAccountResponse updateBankAccount(String id, BankAccountRequest request, UserPrincipal principal);

    void deleteBankAccount(String id, UserPrincipal principal);

    void setDefaultBankAccount(String id, UserPrincipal principal);
}
